### PR TITLE
Update name_id/name checks for empty IDs and empty strings

### PR DESCRIPTION
**Pull request by daniel-j-h**
Here's a failing cucumber scenario where we get a `continue right` when there are no name tags but the ref tag changes. My hunch is the underlying issue is https://github.com/Project-OSRM/osrm-backend/issues/4642 and the `continue` type comes from:

https://github.com/Project-OSRM/osrm-backend/blob/2ae63619a6dbe605b9f07a66daeb3bcbd6e0c708/src/extractor/guidance/intersection_handler.cpp#L81-L89

Next actions:
- Fix https://github.com/Project-OSRM/osrm-backend/issues/4642

### CODE REVIEW

**Comment by karenzshea** [comment]
@daniel-j-h can you review here?

File: include/engine/guidance/collapse_scenario_detection.hpp [hunk 1]
```diff
@@ -36,7 +36,10 @@ bool isStaggeredIntersection(const RouteStepIterator step_prior_to_intersection,
 // a - > x
 bool isUTurn(const RouteStepIterator step_prior_to_intersection,
              const RouteStepIterator step_entering_intersection,
-             const RouteStepIterator step_leaving_intersection);
+             const RouteStepIterator step_leaving_intersection,
+             const std::string &step_prior_name,
+             const std::string &step_entering_name,
+             const std::string &step_leaving_name);
```
Comment (daniel-j-h) [inline include/engine/guidance/collapse_scenario_detection.hpp]: Why do we need the strings here? Aren't they in the route steps already?

File: src/engine/guidance/collapse_turns.cpp [hunk 2]
```diff
@@ -371,6 +371,9 @@ RouteSteps collapseTurnInstructions(RouteSteps steps)
             break;
 
         const auto previous_step = findPreviousTurn(current_step);
+        const auto previous_step_name = facade.GetNameForID(previous_step->name_id).to_string();
+        const auto current_step_name = facade.GetNameForID(current_step->name_id).to_string();
+        const auto next_step_name = facade.GetNameForID(next_step->name_id).to_string();
```
Comment (daniel-j-h) [inline src/engine/guidance/collapse_turns.cpp]: Do we really need a string copy here or is a string view good enough?

(the `.to_string()` makes a copy of the internal memory used to represent a string)

File: include/engine/guidance/collapse_scenario_detection.hpp (hunk 1 above)
Comment (karenzshea) [reply include/engine/guidance/collapse_scenario_detection.hpp]: oo good catch, did not see that this was already available.

File: src/engine/guidance/collapse_turns.cpp (hunk 2 above)
Comment (karenzshea) [reply src/engine/guidance/collapse_turns.cpp]: this can be removed since way names are already available through RouteSteps

File: src/engine/guidance/collapse_scenario_detection.cpp [hunk 3]
```diff
@@ -164,6 +163,9 @@ bool isStaggeredIntersection(const RouteStepIterator step_prior_to_intersection,
 bool isUTurn(const RouteStepIterator step_prior_to_intersection,
              const RouteStepIterator step_entering_intersection,
              const RouteStepIterator step_leaving_intersection)
+//             const std::string &step_prior_name,
+//             const std::string &step_entering_name,
+//             const std::string &step_leaving_name)
```
Comment (daniel-j-h) [inline src/engine/guidance/collapse_scenario_detection.cpp]: Remove

File: src/engine/guidance/collapse_scenario_detection.cpp [hunk 4]
```diff
@@ -41,15 +41,14 @@ bool noIntermediaryIntersections(const RouteStep &step)
 }
 
 // Link roads, as far as we are concerned, are short unnamed segments between to named segments.
-bool isLinkroad(const RouteStep &pre_link_step,
-                const RouteStep &link_step,
-                const RouteStep &post_link_step)
+bool isLinkRoad(const RouteStep &link_step,
+                const std::string &pre_link_step_name,
+                const std::string &post_link_step_name)
```
Comment (daniel-j-h) [inline src/engine/guidance/collapse_scenario_detection.cpp]: Why the change here?

File: src/engine/guidance/collapse_turns.cpp [hunk 5]
```diff
@@ -463,6 +463,7 @@ RouteSteps collapseTurnInstructions(RouteSteps steps)
         if (!hasWaypointType(*previous_step))
         {
             const auto far_back_step = findPreviousTurn(previous_step);
+            const auto far_back_step_name = facade.GetNameForID(far_back_step->name_id).to_string();
```
Comment (daniel-j-h) [inline src/engine/guidance/collapse_turns.cpp]: Unused, can this be removed, and then can the facade also be removed from the function signature?

File: src/extractor/guidance/intersection_handler.cpp [hunk 6]
```diff
@@ -71,17 +71,19 @@ TurnType::Enum IntersectionHandler::findBasicTurnType(const EdgeID via_edge,
...
-    const auto &in_name =
+    const auto &in_name_id =
         node_data_container.GetAnnotation(node_based_graph.GetEdgeData(via_edge).annotation_data)
             .name_id;
-    const auto &out_name =
+    const auto &out_name_id =
         node_data_container.GetAnnotation(node_based_graph.GetEdgeData(road.eid).annotation_data)
             .name_id;
+    const auto &in_name = name_table.GetNameForID(in_name_id).to_string();
+    const auto &out_name = name_table.GetNameForID(out_name_id).to_string();
```
Comment (daniel-j-h) [inline src/extractor/guidance/intersection_handler.cpp]: Do we have to make two string copies here? Or can we already check the string view for empty?

File: src/extractor/guidance/mergable_road_detector.cpp [hunk 7]
```diff
@@ -32,15 +32,17 @@ inline auto makeCheckRoadForName(const NameID name_id,
...
         const MergableRoadDetector::MergableRoadData &road) {
         // since we filter here, we don't want any other name than the one we are looking for
-        const auto road_name =
+        const auto road_name_id =
             node_data_container
                 .GetAnnotation(node_based_graph.GetEdgeData(road.eid).annotation_data)
                 .name_id;
-        if (name_id == EMPTY_NAMEID || road_name == EMPTY_NAMEID)
+        const auto road_name = name_table.GetNameForID(road_name_id).to_string();
+        if (name_id == EMPTY_NAMEID || road_name.empty())
```
Comment (daniel-j-h) [inline src/extractor/guidance/mergable_road_detector.cpp]: I don't understand this - why to we
- receive the name_id
- use the name_id to fetch the road name
- and only then check if the name_id is valid?

what will happen if the name id is invalid and we call GetNameForID on it?

File: src/extractor/guidance/mergable_road_detector.cpp [hunk 8]
```diff
@@ -465,16 +467,19 @@ bool MergableRoadDetector::IsTrafficIsland(const NodeID intersection_node,
...
         const auto has_required_name = [this, required_name_id](const auto edge_id) {
-            const auto road_name =
+            const auto road_name_id =
                 node_data_container
                     .GetAnnotation(node_based_graph.GetEdgeData(edge_id).annotation_data)
                     .name_id;
-            if (required_name_id == EMPTY_NAMEID || road_name == EMPTY_NAMEID)
+            const auto &road_name = name_table.GetNameForID(road_name_id).to_string();
+            const auto &required_name = name_table.GetNameForID(required_name_id).to_string();
+            if (required_name_id == EMPTY_NAMEID || road_name_id == EMPTY_NAMEID ||
```
Comment (daniel-j-h) [inline src/extractor/guidance/mergable_road_detector.cpp]: Same here

File: src/extractor/guidance/motorway_handler.cpp [hunk 9]
```diff
@@ -379,11 +379,15 @@ Intersection MotorwayHandler::fromRamp(const EdgeID via_eid, Intersection inters
             //
             //     7         1
             //          0
+            const auto &first_intersection_name =
+                name_table.GetNameForID(first_intersection_data.name_id).to_string();
+            const auto &second_intersection_name =
+                name_table.GetNameForID(second_intersection_data.name_id).to_string();
```
Comment (daniel-j-h) [inline src/extractor/guidance/motorway_handler.cpp]: And here - what happens if name id is invalid

File: src/extractor/guidance/roundabout_handler.cpp [hunk 10]
```diff
@@ -295,7 +295,8 @@ RoundaboutType RoundaboutHandler::getRoundaboutType(const NodeID nid) const
                     return SPECIAL_EDGEID;
                 }
 
-                if (EMPTY_NAMEID != edge_data.name_id)
+                const auto &edge_name = name_table.GetNameForID(edge_data.name_id).to_string();
+                if (!edge_name.empty())
```
Comment (daniel-j-h) [inline src/extractor/guidance/roundabout_handler.cpp]: here

File: src/extractor/guidance/sliproad_handler.cpp [hunk 11]
```diff
@@ -474,7 +474,8 @@ operator()(const NodeID /*nid*/, const EdgeID source_edge_id, Intersection inter
 
             // Name mismatch: check roads at `c` and `d` for same name
             const auto name_mismatch = [&](const NameID road_name_id) {
-                const auto unnamed = road_name_id == EMPTY_NAMEID;
+                const auto &road_name = name_table.GetNameForID(road_name_id).to_string();
```
Comment (daniel-j-h) [inline src/extractor/guidance/sliproad_handler.cpp]: and here

File: src/extractor/guidance/sliproad_handler.cpp [hunk 12]
```diff
@@ -501,9 +502,13 @@ operator()(const NodeID /*nid*/, const EdgeID source_edge_id, Intersection inter
                     .name_id;
             const auto &sliproad_annotation =
                 node_data_container.GetAnnotation(sliproad_edge_data.annotation_data);
+            const auto &sliproad_name =
+                name_table.GetNameForID(sliproad_annotation.name_id).to_string();
```
Comment (daniel-j-h) [inline src/extractor/guidance/sliproad_handler.cpp]: and here

File: src/extractor/guidance/turn_handler.cpp [hunk 13]
```diff
@@ -199,7 +199,8 @@ bool TurnHandler::isObviousOfTwo(const EdgeID via_edge,
 
     const bool turn_is_perfectly_straight =
         angularDeviation(road.angle, STRAIGHT_ANGLE) < std::numeric_limits<double>::epsilon();
-    if (via_data.name_id != EMPTY_NAMEID)
+    const auto &via_name = name_table.GetNameForID(via_data.name_id).to_string();
```
Comment (daniel-j-h) [inline src/extractor/guidance/turn_handler.cpp]: and here

File: src/extractor/guidance/intersection_handler.cpp (hunk 6 above)
Comment (karenzshea) [reply src/extractor/guidance/intersection_handler.cpp]: good idea!

File: src/extractor/guidance/intersection_handler.cpp [hunk 14]
```diff
@@ -71,17 +71,19 @@ TurnType::Enum IntersectionHandler::findBasicTurnType(const EdgeID via_edge,
...
-    const auto &in_name =
+    const auto &in_name_id =
         node_data_container.GetAnnotation(node_based_graph.GetEdgeData(via_edge).annotation_data)
             .name_id;
-    const auto &out_name =
+    const auto &out_name_id =
         node_data_container.GetAnnotation(node_based_graph.GetEdgeData(road.eid).annotation_data)
             .name_id;
+    const auto &in_name_empty = name_table.GetNameForID(in_name_id).to_string().empty();
+    const auto &out_name_empty = name_table.GetNameForID(out_name_id).to_string().empty();
```
Comment (daniel-j-h) [inline src/extractor/guidance/intersection_handler.cpp]: `GetXForID` returns

http://www.boost.org/doc/libs/1_60_0/libs/utility/doc/html/string_ref.html

a zero-copy view onto the data. If you call `.to_string()` on it you make a copy of the data and allocate memory for that. But we only want to know if the data is empty. Check the string ref docs, there is already an `empty` function on the string ref.

File: src/extractor/guidance/mergable_road_detector.cpp [hunk 15]
```diff
@@ -32,15 +34,20 @@ inline auto makeCheckRoadForName(const NameID name_id,
...
-        const auto road_name =
+        const auto road_name_id =
             node_data_container
                 .GetAnnotation(node_based_graph.GetEdgeData(road.eid).annotation_data)
                 .name_id;
-        if (name_id == EMPTY_NAMEID || road_name == EMPTY_NAMEID)
+        if (name_id == EMPTY_NAMEID || road_name_id == EMPTY_NAMEID)
+            return true;
+        const auto road_name_empty = name_table.GetNameForID(road_name_id).to_string().empty();
+        const auto in_name_empty = name_table.GetNameForID(name_id).to_string().empty();
```
Comment (daniel-j-h) [inline src/extractor/guidance/mergable_road_detector.cpp]: same here

File: src/extractor/guidance/mergable_road_detector.cpp [hunk 16]
```diff
@@ -463,18 +470,25 @@ bool MergableRoadDetector::IsTrafficIsland(const NodeID intersection_node,
...
+            const auto road_name_id =
                 node_data_container
                     .GetAnnotation(node_based_graph.GetEdgeData(edge_id).annotation_data)
                     .name_id;
-            if (required_name_id == EMPTY_NAMEID || road_name == EMPTY_NAMEID)
+            if (road_name_id == EMPTY_NAMEID)
+                return false;
+            const auto &road_name_empty = name_table.GetNameForID(road_name_id).to_string().empty();
+            const auto &required_name_empty =
+                name_table.GetNameForID(required_name_id).to_string().empty();
```
Comment (daniel-j-h) [inline src/extractor/guidance/mergable_road_detector.cpp]: same here

File: src/extractor/guidance/sliproad_handler.cpp [hunk 17]
```diff
@@ -474,7 +474,10 @@ operator()(const NodeID /*nid*/, const EdgeID source_edge_id, Intersection inter
 
             // Name mismatch: check roads at `c` and `d` for same name
             const auto name_mismatch = [&](const NameID road_name_id) {
-                const auto unnamed = road_name_id == EMPTY_NAMEID;
+                const auto unnamed =
+                    road_name_id == EMPTY_NAMEID
+                        ? true
+                        : name_table.GetNameForID(road_name_id).to_string().empty();
```
Comment (daniel-j-h) [inline src/extractor/guidance/sliproad_handler.cpp]: same here

File: src/extractor/guidance/sliproad_handler.cpp [hunk 18]
```diff
@@ -501,9 +504,15 @@ operator()(const NodeID /*nid*/, const EdgeID source_edge_id, Intersection inter
                     .name_id;
             const auto &sliproad_annotation =
                 node_data_container.GetAnnotation(sliproad_edge_data.annotation_data);
+            const auto &sliproad_name_empty =
+                sliproad_annotation.name_id == EMPTY_NAMEID
+                    ? true
+                    : name_table.GetNameForID(sliproad_annotation.name_id).to_string().empty();
+            const auto &main_road_name = name_table.GetNameForID(main_road_name_id).to_string();
+            const auto &candidate_road_name =
+                name_table.GetNameForID(candidate_data.name_id).to_string();
```
Comment (daniel-j-h) [inline src/extractor/guidance/sliproad_handler.cpp]: same here

File: src/extractor/guidance/turn_handler.cpp [hunk 19]
```diff
@@ -199,7 +199,11 @@ bool TurnHandler::isObviousOfTwo(const EdgeID via_edge,
 
     const bool turn_is_perfectly_straight =
         angularDeviation(road.angle, STRAIGHT_ANGLE) < std::numeric_limits<double>::epsilon();
-    if (via_data.name_id != EMPTY_NAMEID)
+    const auto &via_name_empty =
+        via_data.name_id == EMPTY_NAMEID
+            ? true
+            : name_table.GetNameForID(via_data.name_id).to_string().empty();
```
Comment (daniel-j-h) [inline src/extractor/guidance/turn_handler.cpp]: same here

File: src/extractor/guidance/intersection_handler.cpp (hunk 14 above)
Comment (karenzshea) [reply src/extractor/guidance/intersection_handler.cpp]: 🙇‍♀️

File: src/extractor/guidance/motorway_handler.cpp [hunk 20]
```diff
@@ -361,107 +361,120 @@ Intersection MotorwayHandler::fromRamp(const EdgeID via_eid, Intersection inters
...
+                //     5         3
+                //
+                //   6              2
+                //
+                //     7         1
+                //          0
+                const auto &first_intersection_name =
+                    name_table.GetNameForID(first_intersection_data.name_id).to_string();
+                const auto &second_intersection_name =
+                    name_table.GetNameForID(second_intersection_data.name_id).to_string();
```
Comment (daniel-j-h) [inline src/extractor/guidance/motorway_handler.cpp]: Copy needed here?

File: src/extractor/guidance/roundabout_handler.cpp [hunk 21]
```diff
@@ -295,18 +295,22 @@ RoundaboutType RoundaboutHandler::getRoundaboutType(const NodeID nid) const
                     return SPECIAL_EDGEID;
                 }
 
-                if (EMPTY_NAMEID != edge_data.name_id)
+                if (edge_data.name_id != EMPTY_NAMEID)
                 {
+                    const auto &edge_name = name_table.GetNameForID(edge_data.name_id).to_string();
+                    if (!edge_name.empty())
```
Comment (daniel-j-h) [inline src/extractor/guidance/roundabout_handler.cpp]: Copy needed here for the empty check?

File: src/extractor/guidance/sliproad_handler.cpp [hunk 22]
```diff
@@ -499,11 +501,22 @@ operator()(const NodeID /*nid*/, const EdgeID source_edge_id, Intersection inter
                 node_data_container
                     .GetAnnotation(node_based_graph.GetEdgeData(main_road.eid).annotation_data)
                     .name_id;
+            const auto main_road_name_empty =
+                main_road_name_id == EMPTY_NAMEID
+                    ? true
+                    : name_table.GetNameForID(main_road_name_id).empty();
```
Comment (daniel-j-h) [inline src/extractor/guidance/sliproad_handler.cpp]: Here and below: you can simplify to

    const auto empty = name_id == EMPTY || GetName(name_id).empty()

File: src/extractor/guidance/turn_handler.cpp [hunk 23]
```diff
@@ -199,7 +199,9 @@ bool TurnHandler::isObviousOfTwo(const EdgeID via_edge,
 
     const bool turn_is_perfectly_straight =
         angularDeviation(road.angle, STRAIGHT_ANGLE) < std::numeric_limits<double>::epsilon();
-    if (via_data.name_id != EMPTY_NAMEID)
+    const auto &via_name_empty =
+        via_data.name_id == EMPTY_NAMEID ? true : name_table.GetNameForID(via_data.name_id).empty();
```
Comment (daniel-j-h) [inline src/extractor/guidance/turn_handler.cpp]: same

**Comment by karenzshea** [comment]
Cool! I used https://github.com/mapbox/osrm-guidance-differ with a few modifications to compare current master against this branch and I saw some good changes from `continue` to `turn` maneuver types. On master, we sometimes announce a `continue` onto streets that are missing names, and now we announce a `turn` even if the street is unnamed. Mostly saw this on sliproads.
image
_Previously we announced a maneuver onto the unnamed sliproad as a `continue slight right`, is now a `turn slight right`_

Same here, 
image

[Map linked here](https://gist.github.com/5e42b62f1c4fa1881230898157379464)
